presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp

```diff
@@ -1871,15 +1871,20 @@ core::PlanNodePtr VeloxQueryPlanConverterBase::toVeloxQueryPlan(
     const std::shared_ptr<protocol::TableWriteInfo>& tableWriteInfo,
     const protocol::TaskId& taskId) {
   auto joinType = toJoinType(node->type);
-
-  if (node->criteria.empty() && core::isInnerJoin(joinType) && !node->filter) {
-    return std::make_shared<core::NestedLoopJoinNode>(
-        node->id,
-        toVeloxQueryPlan(node->left, tableWriteInfo, taskId),
-        toVeloxQueryPlan(node->right, tableWriteInfo, taskId),
-        toRowType(node->outputVariables));
+  if (node->criteria.empty()) {
+    auto isNestedJoinType = core::isInnerJoin(joinType) |
+        core::isLeftJoin(joinType) | core::isRightJoin(joinType) |
+        core::isFullJoin(joinType);
+    if (isNestedJoinType) {
+      return std::make_shared<core::NestedLoopJoinNode>(
+          node->id,
+          joinType,
+          node->filter ? exprConverter_.toVeloxExpr(*node->filter) : nullptr,
+          toVeloxQueryPlan(node->left, tableWriteInfo, taskId),
+          toVeloxQueryPlan(node->right, tableWriteInfo, taskId),
+          toRowType(node->outputVariables));
+    }
```

Comment:
HashJoin require criteria. If there are no criteria and NestedLoopJoinNode cannot be used, then should we error at this point ?

Reply:
There are checks in HashJoin that fails anyway. Adding here may seem redundant.

Reply:
This is an earlier fail point. 

The HashJoin failure is mis-leading because it gives the impression that the query was expected to use HashJoin and HashJoin doesn't have the features to support it. Hash-Join it is just not expected to handle criteria and some other solution is needed, which will be added here eventually.